### ISSUE TITLE
Strange behavior of typeclasses eauto wrt "shelve" hints

### ISSUE THREAD

**Issue by Armael** [coq/coq#6195]
#### Version
Tested with coq 8.6 and 8.7.

#### Description of the problem
The snippet below displays the following issue: `typeclasses eauto with mymonotonic` is able to prove a given goal, using the `mymonotonic` hints base, containing a single hint of cost 0. However, adding a `shelve` hint with very high priority to the hint base seemingly makes `typeclasses eauto` skip the previous hint (that was able to finish the proof), and use the `shelve` hint instead. This is an issue since I want the shelve hint to be considered last (as indicated by the high priority), only trying all the other hints.

Interestingly, I failed to reduce this snippet further -- that is, the bug would not manifest after:
- explicitly providing the `A` and `B` arguments instead of implicitly;
- hardwiring either `A` or `B` to be `nat` in `monotonic` (leaving the other one implicit);
- getting rid of `foo`
- doing `intro` before calling `typeclasses eauto`

```coq
Parameter monotonic : forall {A B} (leA : A -> A -> Prop) (leB : B -> B -> Prop),
  (A -> B) -> Prop.

Hypothesis monotonic_cst : forall A B (leA : A -> A -> Prop) (leB : B -> B -> Prop),
  forall (b:B), monotonic leA leB (fun _ : A => b).

Hint Extern 0 (monotonic _ _ (fun _ => ?x)) =>
  simple apply monotonic_cst : mymonotonic.

Parameter (foo : nat -> nat).

Goal (forall a, monotonic le le (fun _ => foo a)).
  typeclasses eauto with mymonotonic.
Qed.


Hint Extern 999999 (monotonic _ _ _) => shelve : mymonotonic.

Goal (forall a, monotonic le le (fun _ => foo a)).
  unshelve (typeclasses eauto with mymonotonic).
  admit.
Admitted.
```

**Comment by Zimmi48** [comment]
The issue is that `typeclasses eauto` only tries `intro` when everything else has failed. Here it `shelve` solves the goal, so it never tries `intro`. This might be a design bug but I don't know what to think about this.

**Comment by Zimmi48** [comment]
Note that this is yet another instance of incompatibility between `eauto` (which introduces everything systematically) and `typeclasses eauto` (which does it only in case of failure).

**Comment by mattam82** [comment]
One thing you could do here is check that no product is here in the Hint Extern then, right?

**Comment by Zimmi48** [comment]
> One thing you could do here is check that no product is here in the Hint Extern then, right?

That's a good suggestion but how do you express such a negative pattern?

**Comment by mattam82** [comment]
Indeed the Hint Extern pattern matches the conclusion of the products in the goal, not the whole goal. I personally find this to be a design bug.

@Zimmi48 like this:


Hint Extern 999999 (monotonic _ _ _) => match goal with |- _ -> _ => fail 1
                                                   | _ => shelve end : mymonotonic.

Goal (forall a, monotonic le le (fun _ => foo a)).
  unshelve (typeclasses eauto with mymonotonic).
  admit.
Admitted.

**Comment by mattam82** [comment]
TODO: check that the doc actually explains the Hint Extern behavior

**Comment by Armael** [comment]
Thanks for the explanations. 

As an alternative to the trick you described, would it make sense to have instead:
```
Hint Extern 1 => match goal with |- forall _, _ => intro end : mymonotonic.
Hint Extern 99999 (monotonic _ _ _) => shelve : mymonotonic.
```
as a workaround to simulate the behavior of `eauto`?

**Comment by Zimmi48** [comment]
Putting `intro` in a `Hint Extern` is something that I've tried earlier. Until I realized that it was the source of a large performance loss in case of failure of the proof search. Because the introduction rule that `typeclasses eauto` applies in case of failure is still there, this makes the search space grow exponentially on the number of possible introduction steps.

**Comment by mrhaandi** [comment]
This seems to be fixed now (I tried Coq 8.15.2; in 8.13.2 it was still broken).
